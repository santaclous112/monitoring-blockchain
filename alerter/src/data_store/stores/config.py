import json
import logging
from datetime import datetime
from typing import Dict

import pika.exceptions

from src.message_broker.rabbitmq.rabbitmq_api import RabbitMQApi
from src.data_store.redis.store_keys import Keys
from src.data_store.stores.store import Store
from src.utils.constants import (CONFIG_EXCHANGE, HEALTH_CHECK_EXCHANGE,
                                 STORE_CONFIGS_QUEUE_NAME,
                                 STORE_CONFIGS_ROUTING_KEY_CHAINS)
from src.utils.exceptions import (ReceivedUnexpectedDataException,
                                  MessageWasNotDeliveredException)


class ConfigStore(Store):
    def __init__(self, name: str, logger: logging.Logger,
                 rabbitmq: RabbitMQApi) -> None:
        super().__init__(name, logger, rabbitmq)

    def _initialise_rabbitmq(self) -> None:
        """
        Initialise the necessary data for rabbitmq to be able to reach the data
        store as well as appropriately communicate with it.

        Creates a config exchange of type `topic`
        Declares a queue named `github_store_queue` and binds it to the store
        exchange with a routing key `github` meaning anything
        coming from the transformer with regards to github updates will be
        received here.

        The HEALTH_CHECK_EXCHANGE is also declared so that whenever a successful
        store round occurs, a heartbeat is sent
        """
        self.rabbitmq.connect_till_successful()
        self.logger.info("Creating exchange '%s'", CONFIG_EXCHANGE)
        self.rabbitmq.exchange_declare(CONFIG_EXCHANGE, 'topic', False, True,
                                       False, False)
        self.logger.info("Creating queue '%s'", STORE_CONFIGS_QUEUE_NAME)
        self.rabbitmq.queue_declare(STORE_CONFIGS_QUEUE_NAME, False, True,
                                    False, False)
        self.logger.info("Binding queue '%s' to exchange '%s' with routing "
                         "key '%s'", STORE_CONFIGS_QUEUE_NAME,
                         CONFIG_EXCHANGE, '#')
        self.rabbitmq.queue_bind(queue=STORE_CONFIGS_QUEUE_NAME,
                                 exchange=CONFIG_EXCHANGE,
                                 routing_key='#')
        self.logger.info("Creating '%s' exchange", HEALTH_CHECK_EXCHANGE)
        self.rabbitmq.exchange_declare(HEALTH_CHECK_EXCHANGE, 'topic', False,
                                       True, False, False)

    def _listen_for_data(self) -> None:
        self.rabbitmq.basic_consume(queue=STORE_CONFIGS_QUEUE_NAME,
                                    on_message_callback=self._process_data,
                                    auto_ack=False,
                                    exclusive=False, consumer_tag=None)
        self.rabbitmq.start_consuming()
        self.logger.info("Setting delivery confirmation on RabbitMQ channel")
        self.rabbitmq.confirm_delivery()

    def _process_data(self,
                      ch: pika.adapters.blocking_connection.BlockingChannel,
                      method: pika.spec.Basic.Deliver,
                      properties: pika.spec.BasicProperties,
                      body: bytes) -> None:
        """
        Processes the data being received, from the queue. This data will be
        stored in Redis as required. If successful, a heartbeat will be sent.
        """
        config_data = json.loads(body.decode())

        if 'DEFAULT' in config_data:
            del config_data['DEFAULT']

        self.logger.debug("Received %s. Now processing this data.", config_data)

        processing_error = False
        try:
            self._process_redis_store(method.routing_key, config_data)
        except KeyError as e:
            self.logger.error("Error when parsing %s.", config_data)
            self.logger.exception(e)
            processing_error = True
        except ReceivedUnexpectedDataException as e:
            self.logger.error("Error when processing %s", config_data)
            self.logger.exception(e)
            processing_error = True
        except Exception as e:
            self.logger.exception(e)
            processing_error = True

        self.rabbitmq.basic_ack(method.delivery_tag, False)

        # Send a heartbeat only if there were no errors
        if not processing_error:
            try:
                heartbeat = {
                    'component_name': self.name,
                    'timestamp': datetime.now().timestamp()
                }
                self._send_heartbeat(heartbeat)
            except MessageWasNotDeliveredException as e:
                self.logger.exception(e)
            except Exception as e:
                # For any other exception raise it.
                raise e

    def _process_redis_store(self, routing_key: str, data: Dict) -> None:
        if 'general' in routing_key[0]:
            if 'systems_config' in routing_key[1]:
                self._process_redis_config_system(routing_key[0], data)
            elif 'alerts_config' in routing_key[1]:
                self._process_redis_config_alerts(routing_key[0], data)
            elif 'repos_config' in routing_key[1]:
                self._process_redis_config_repos(routing_key[0], data)
        elif 'chains' in routing_key[0]:
            if 'nodes_config' in routing_key[1]:
                self._process_redis_config_nodes(routing_key[0], data)
            elif 'alerts_config' in routing_key[1]:
                self._process_redis_config_alerts(routing_key[0], data)
            elif 'repos_config' in routing_key[1]:
                self._process_redis_config_repos(routing_key[0], data)
        elif 'channels' in routing_key[0]:
            if 'opsgenie_config' in routing_key[1]:
                
            elif 'pagerduty_config' in routing_key[1]:
            elif 'telegram_config' in routing_key[1]:
            elif 'email_config' in routing_key[1]:

                self._process_redis_config_channel(data)

    def _process_redis_config_system(self, parent_id: str, data: Dict) -> None:
        for key, value in data.items():
            self.logger.debug(
                "Saving %s state: config=%s", key, value
            )

            self.redis.hset(Keys.get_hash_parent(parent_id), {
                Keys.get_config_chain_system(key): value
            })

    def _process_redis_config_alerts(self, parent_id: str, data: Dict) -> None:
        for key, value in data.items():
            self.logger.debug(
                "Saving %s state: config=%s", key, value
            )

            self.redis.hset(Keys.get_hash_parent(parent_id), {
                Keys.get_config_chain_alerts(key): value
            })

    def _process_redis_config_repos(self, parent_id: str, data: Dict) -> None:
        for key, value in data.items():
            self.logger.debug(
                "Saving %s state: config=%s", key, value
            )

            self.redis.hset(Keys.get_hash_parent(parent_id), {
                Keys.get_config_chain_repo(key): value
            })

    def _process_redis_config_channel(self, parent_id: str, data: Dict) -> None:
        for key, value in data.items():
            self.logger.debug(
                "Saving %s state: config=%s", key, value
            )

            self.redis.hset(Keys.get_hash_parent(parent_id), {
                Keys.get_config_chain_repo(key): value
            })

    def _process_redis_config_nodes(self, parent_id: str, data: Dict) -> None:
        for key, value in data.items():
            self.logger.debug(
                "Saving %s state: config=%s", key, value
            )

            self.redis.hset(Keys.get_hash_parent(parent_id), {
                Keys.get_config_chain_node(key): value
            })
