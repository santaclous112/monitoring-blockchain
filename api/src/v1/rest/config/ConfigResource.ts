import express from "express";
import {ObjectID} from "mongodb";
import mongoose, {Document} from "mongoose";
import {Config} from "../../../../../entities/ts/Config";
import {GenericDocument} from "../../../constant/mongoose";
import {
    CouldNotRemoveDataFromDB,
    CouldNotRetrieveDataFromDB,
    CouldNotSaveDataToDB,
    DuplicateWarning,
    InvalidIDError,
    MissingParameterWarning,
    NotFoundWarning,
    ValidationDataError
} from "../../../constant/server.feedback";
import {MongooseUtil} from "../../../util/MongooseUtil";
import {ResponseError, ResponseNoContent, ResponseSuccess}
    from "../../entity/io/ResponseData";
import {ConfigModel} from "../../entity/model/ConfigModel";
import {ConfigRepository} from "../../entity/repository/ConfigRepository";
import {BaseChainRepository}
    from "../../entity/repository/BaseChainRepository";
import {ObjectUtil} from "../../../util/ObjectUtil";
import {EmailRepository} from "../../entity/repository/ChannelRepository";
import {ConfigOldModel} from "../../entity/model/ConfigOldModel";

/**
 * Resource Controller for Panic Configuration
 */
export class ConfigResource {
    private configRepo: ConfigRepository = null;
    private baseChainRepo: BaseChainRepository = null;
    private channelRepo: EmailRepository = null;

    constructor() {
        this.configRepo = new ConfigRepository();
        this.baseChainRepo = new BaseChainRepository();
        this.channelRepo = new EmailRepository();
    }

    /**
     * Get a list of Configurations from Database in config collection
     *
     * @param req Request from Express
     * @param res Response from Express
     */
    public async getAll(req: express.Request,
                        res: express.Response): Promise<void> {
        try {
            const configs = await this.configRepo.findAll();

            const response = new ResponseSuccess<Config[]>(res);
            response.send(configs);
        } catch (err: any) {
            console.error(err);
            const error = new CouldNotRetrieveDataFromDB();
            const response = new ResponseError(res, error);
            response.send();
        }
    }

    /**
     * Get configuration by id from Database in config collection
     *
     * @param req Request from Express
     * @param res Response from Express
     */
    public async getItem(req: express.Request,
        res: express.Response): Promise<void> {
        try {
            const responseError = this.isMissingParam(req, res);
            if (responseError instanceof ResponseError) {
                responseError.send();
                return;
            }

            const config = await this.getConfigById(res, req.params.id);
            if (config instanceof ResponseError) {
                config.send();
                return;
            }

            const response = new ResponseSuccess<Config>(res);
            response.send(config);
        } catch (err: any) {
            console.error(err);
            const error = new CouldNotRetrieveDataFromDB();
            const response = new ResponseError(res, error);
            response.send();
        }
    }

    /**
     * Create a new Configuration on the Database in the config collection
     *
     * @param req Request from Express
     * @param res Response from Express
     */
    public async create(req: express.Request,
        res: express.Response): Promise<void> {
        try {
            const invalidFields = [
                'threshold_alerts', 'severity_alerts', 'time_window_alerts'
            ];
            
            let config: Config = MongooseUtil.merge(
                new Config(), req.body, invalidFields);

            // subChain name duplication validation
            let duplicate = await this.configRepo.isDuplicateSubChain(config);
            if (duplicate) {
                const response = new ResponseError(
                    res, new DuplicateWarning('subChain.name'));
                response.send();
                return;
            }

            let baseChain = await this.baseChainRepo.findOneByIdAndDeepPopulate(
                req.body.baseChain.id || req.body.baseChain,
                ['severity_alerts', 'threshold_alerts', 'time_window_alerts']);

            baseChain = ObjectUtil.snakeToCamel(baseChain.toJSON());

            // populate alerts with baseChains alerts
            config.thresholdAlerts = baseChain.thresholdAlerts;
            config.severityAlerts = baseChain.severityAlerts;
            config.timeWindowAlerts = baseChain.timeWindowAlerts;

            const model: Document<Config> = new ConfigModel(config.toJSON());
            await this.save(res, model);
        } catch (err: any) {
            console.error(err);
            const error = new CouldNotSaveDataToDB();
            const response = new ResponseError(res, error);
            response.send();
        }
    }

    /**
     * Updates an existing configuration on the Database in the config
     * collection
     *
     * @param req Request from Express
     * @param res Response from Express
     */
    public async update(req: express.Request,
                        res: express.Response): Promise<void> {
        try {
            const responseError = this.isMissingParam(req, res);
            if (responseError instanceof ResponseError) {
                responseError.send();
                return;
            }

            // to avoid updating baseChain on update
            delete req.body.baseChain;

            // subChain name duplication validation
            let duplicate = await this.configRepo.isDuplicateSubChain({
                ...req.body,
                id: req.params.id
            } as Config);

            if (duplicate) {
                const response = new ResponseError(res,
                    new DuplicateWarning('subChain.name'));
                response.send();
                return;
            }

            const config: Document<Config> = await this.getConfigById(res,
                req.params.id) as any;
            if (config instanceof ResponseError) {
                config.send();
                return;
            }

            await this.createBkp(config);

            const request = ObjectUtil.deepCamelToSnake(req.body);
            const model: Document = MongooseUtil.merge(config, request);

            await this.save(res, model);
        } catch (err: any) {
            console.error(err);
            const error = new CouldNotSaveDataToDB();
            const response = new ResponseError(res, error);
            response.send();
        }
    }

    /**
     * Removes configuration by id on Database in config collection
     *
     * @param req Request from Express
     * @param res Response from Express
     */
    public async remove(req: express.Request,
                        res: express.Response): Promise<void> {
        try {
            const responseError = this.isMissingParam(req, res);
            if (responseError instanceof ResponseError) {
                responseError.send();
                return;
            }

            const config = await this.configRepo.findOneById(req.params.id);
            await this.createBkp(config as any);

            const isDeleted = await this.configRepo
                                        .deleteOneById(req.params.id);
            console.log(isDeleted);                                        
            if (!isDeleted) {
                const response = new ResponseError(res, new NotFoundWarning());
                response.send();
                return;
            }            

            await this.channelRepo.unlinkConfigFromAllChannels(req.params.id);

            const response = new ResponseNoContent(res);
            response.send();
        } catch (err: any) {
            console.error(err);
            const error = new CouldNotRemoveDataFromDB();
            const response = new ResponseError(res, error);
            response.send();
        }
    }

    /**
     * Checks if there are missing params
     *
     * @param req Request from Express
     * @param res Response from Express
     */
    private isMissingParam(req: express.Request,
                           res: express.Response): ResponseError {
        if (!req.params.id) {
            const error = new MissingParameterWarning('id');

            return new ResponseError(res, error);
        }

        return null;
    }

    /**
     * Save document on config collection. If an error occurs, an error response
     * is sent via `res`.
     *
     * @param res Response from Express
     * @param model The config document object
     */
    private async save(res: express.Response, model: Document<Config>): 
        Promise<void> {

        model.set('config_type', new ObjectID(
            GenericDocument.CONFIG_TYPE_SUB_CHAIN)
        );

        const isValid = await MongooseUtil.isValid(model);
        if (!isValid) {
            const error = new ValidationDataError();

            let response = new ResponseError(res, error);
            await response.addMongooseErrors(model);

            const duplicateError = response.messages.find(
                message => message.description.includes('Name duplicated'));
            if (duplicateError) {
                response = new ResponseError(res, new DuplicateWarning(
                    `${duplicateError.name} name`));
            }

            response.send();
            return;
        }

        const doc = await model.save();

        const response = new ResponseSuccess<Document>(res);
        response.send(doc.id);
    }

    /**
     * Get configuration by ID, if not found send a `NotFoundWarning` over a 
     * `ResponseError` instance.
     *
     * @param res Response from Express
     * @param id Configuration ID
     * @returns a promise containing either a `Config` object or a 
     *  `ResponseError` instance
     */
    private async getConfigById(res: express.Response, id: string): 
        Promise<Config | ResponseError> {
        if (!mongoose.Types.ObjectId.isValid(id)) {
            const error = new InvalidIDError(id);
            return new ResponseError(res, error);
        }

        const config = await this.configRepo.findOneById(id);
        if (!config) {
            return new ResponseError(res, new NotFoundWarning());
        }

        return config;
    }

    /**
     * Create a bkp to comparison on alerter
     * @param config  Current official config
     */
    private async createBkp(config : Document<Config>) : Promise<void> {        
        await ConfigOldModel.deleteOne({ _id: config._id });
        const oldModel : Document<Config> = new ConfigOldModel(
            config.toObject()
        );
        await oldModel.save();
    }
}
