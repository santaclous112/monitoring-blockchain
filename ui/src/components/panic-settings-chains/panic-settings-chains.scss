@import '../scss/mixins/heading.scss';
@import '../scss/mixins/buttons.scss';
@import '../scss/mixins/centralized-container.scss';
@import '../scss/mixins/data-tables.scss';
@import '../scss/mixins/nav-buttons.scss';

.panic-settings-chains {

  &__container {
    padding-left: 5%;
    padding-right: 5%;
  }

  &__crud-data-table {
    @include crud-data-table();
  }

  &__title {
    padding-top: 15px;
    @include heading-title();
  }

  &__add-sub-chain-button {
    @include add-button();
  }
}

// Menu config modal
.panic-settings-config-menu-modal {
  --height: 345px !important;

  svc-content-container {
    padding: 10px;
  }

  svc-button {
    @include button(190px);
    @include nav-buttons(0 0 5px 0);
  }
}