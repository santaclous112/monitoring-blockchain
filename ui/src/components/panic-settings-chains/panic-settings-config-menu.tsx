import { dismissModal } from '@simply-vc/uikit';
import { h, Component, Prop, Host } from '@stencil/core';
import { Config } from '../../../../entities/ts/Config';

@Component({
    tag: "panic-settings-config-menu"
})
export class PaniSettingsConfigMenu {

    /**
     * The config being edited.
     */
    @Prop() config: Config;

    /**
     * Checks if the current URL contains `path`.
     * 
     * @param path URL path
     * @returns `true` if the current URL contais `path`, otherwise `false`.
     */
    isDisabled(path: string): boolean {
        return window.location.pathname.indexOf(path) !== -1;
    }

    render() {
        return (
            <Host>
                <svc-header headline={`Choose what to edit for ${this.config.subChain.name}`} menuPosition={"end"}>
                    <ion-buttons slot={"menu"}>
                        <svc-button iconPosition={"icon-only"} iconName={"close"} onClick={() => dismissModal() } />
                    </ion-buttons>
                </svc-header>
                <svc-content-container>
                    <svc-button
                        disabled={this.isDisabled("sub-chain")}
                        iconName={"link"}
                        iconPosition={"start"}
                        color={"secondary"}
                        href={`/settings/edit/sub-chain/${this.config.id}`}
                    >
                        Sub-chain
                    </svc-button>
                    <svc-button
                        disabled={this.isDisabled("channels")}
                        iconName={"call"}
                        iconPosition={"start"}
                        color={"secondary"}
                        href={`/settings/edit/channels/${this.config.id}`}
                    >
                        Channels
                    </svc-button>
                    <svc-button
                        disabled={this.isDisabled("sources")}
                        iconName={"server"}
                        iconPosition={"start"}
                        color={"secondary"}
                        href={`/settings/edit/sources/${this.config.id}`}
                    >
                        Nodes
                    </svc-button>
                    <svc-button
                        disabled={this.isDisabled("repositories")}
                        iconName={"code-slash"}
                        iconPosition={"start"}
                        color={"secondary"}
                        href={`/settings/edit/repositories/${this.config.id}`}
                    >
                        Repositories
                    </svc-button>
                    <svc-button
                        disabled={this.isDisabled("alerts")}
                        iconName={"notifications-circle"}
                        iconPosition={"start"}
                        color={"secondary"}
                        href={`/settings/edit/alerts/${this.config.id}`}
                    >
                        Alerts
                    </svc-button>
                </svc-content-container>
            </Host>
        );
    }
}