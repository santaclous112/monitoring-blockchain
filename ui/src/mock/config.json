[
  {
    "id": "62675e2d8891cd77b87f5b16",
    "status": true,
    "created": "2022-05-02T22:02:46.000Z",
    "modified": null,
    "baseChain": {
      "id": "6265cefcfdb17d641746dced",
      "status": true,
      "name": "Cosmos",
      "value": "cosmos",
      "description": "Cosmos description here...",
      "sources": [
        {
          "id": "6271178cf740ef3847a7d27e",
          "status": true,
          "created": "2022-04-01T00:00:00.000Z",
          "modified": null,
          "name": "cosmos_node_1",
          "value": null,
          "description": null
        }
      ]
    },
    "subChain": {
      "id": "626aeb43980cf43ee7bbf683",
      "status": true,
      "created": "2022-05-02T22:02:46.000Z",
      "modified": null,
      "name": "Akash"
    }
  },
  {
    "id": "62675e2d8891cd77b87f5b166",
    "status": true,
    "created": "2022-05-02T22:02:46.000Z",
    "modified": null,
    "baseChain": {
      "id": "6265cefcfdb17d641746dcedd",
      "status": true,
      "name": "Cosmos",
      "value": "cosmos",
      "description": "Cosmos description here...",
      "sources": [
        {
          "id": "6271178cf740ef3847a7d27e",
          "status": true,
          "created": "2022-04-01T00:00:00.000Z",
          "modified": null,
          "name": "cosmos_node_2",
          "value": null,
          "description": null
        }
      ]
    },
    "subChain": {
      "id": "626aeb43980cf43ee7bbf6833",
      "status": true,
      "created": "2022-05-02T22:02:46.000Z",
      "modified": null,
      "name": "Akash 2"
    }
  }
]