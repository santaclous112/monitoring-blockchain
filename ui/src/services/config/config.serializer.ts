import { Config } from "../../../../entities/ts/Config";

/**
 * Converts "raw data" (JSON response) into {@link ConfigInterface} object and vice-versa.
 */
export class ConfigSerializer {

    unserialize(jsonObject: object): Config {
        return Config.fromJSON(jsonObject, Config);
    }

    serialize(entity: Config): string {
        return JSON.stringify(entity);
    }
}
