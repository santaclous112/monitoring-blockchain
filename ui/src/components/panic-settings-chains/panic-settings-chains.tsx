import {Component, h, Host, Listen, Prop, State} from '@stencil/core';
import {
  DataTableRowsType,
  DataTableRow
} from "@simply-vc/uikit/dist/types/types/datatable";
import {Config} from "../../../../entities/ts/Config";
import {CONFIG_ID_LOCAL_STORAGE_KEY, SUBCHAIN_TABLE_HEADERS} from "../../utils/constants";
import {ConfigService} from '../../services/config/config.service';
import {Router} from "stencil-router-v2";
import {HelperAPI} from "../../utils/helpers";
import {createAlert, createModal} from "@simply-vc/uikit";

@Component({
  tag: 'panic-settings-chains',
  styleUrl: 'panic-settings-chains.scss'
})
export class PanicSettingsChains {

  /**
   * Stencil Router object for page navigation.
   */
  @Prop() router: Router;

  /**
   * List of configurations available to the node operator.
   */
  @State() configs: Config[] = [];

  async componentWillLoad() {
    this.configs = await ConfigService.getInstance().getAll();

    // cleaning configId to avoid side-effects
    localStorage.removeItem(CONFIG_ID_LOCAL_STORAGE_KEY);
  }

  /**
   * Listens to the `editConfig` {@link CustomEvent} to trigger the navigation.
   */
  @Listen("editConfig", {target: "window"})
  async editConfigHandler(e: CustomEvent){
    const id: string = e.detail.id
    localStorage.setItem(CONFIG_ID_LOCAL_STORAGE_KEY, id);
    const config: Config = await ConfigService.getInstance().getByID(id);

    createModal("panic-settings-config-menu", {
      config: config
    },
    {
      cssClass: "panic-settings-config-menu-modal"
    })
  }

  /**
   * Given a Config object ID this function will retrieve the index of the
   * object from `this.configs` matching the ID
   * @param configId The Config object ID
   */
  findConfigIndex(configId: string): number {
    return this.configs.findIndex(config => config.id === configId)
  }

  /**
   * Whenever a deleteConfig is detected, this handler displays a confirmation
   * modal
   * @param event The raised deleteConfig event
   */
  @Listen('deleteConfig', {target: 'window'})
  deleteConfigHandler(event: CustomEvent) {
    const configIndex = this.findConfigIndex(event.detail.id)
    const config = this.configs[configIndex]

    createAlert({
      'header': 'Delete Confirmation',
      'message': `Are you sure you want to delete the configuration for ${config.subChain.name}?`,
      'eventName': 'configDeleteConfirmation',
      'eventData': config
    });
  }

  /**
   * Whenever a response is detected on the config delete confirmation modal
   * this handler does the following:
   *
   * If Deletion Confirmed: Performs an API call to remove the config from the
   *                        database and removes the config from the component
   *                        state
   * If Deletion canceled: Does nothing
   * @param event
   */
  @Listen('configDeleteConfirmation', {target: 'window'})
  async handleConfigDeleteConfirmation(event: CustomEvent) {
    const deletionConfirmed = event.detail.confirmed;
    const configRemove = event.detail.data;

    if (deletionConfirmed){
      const removed: boolean = await ConfigService.getInstance().delete(configRemove.id);

      if (removed) {
        this.configs = this.configs.filter(config => config.id !== configRemove.id);
        HelperAPI.raiseToast("Chain configuration removed successfully!");
      } else {
        HelperAPI.raiseToast("Could not delete configuration from database!", 2000, "danger");
      }
    }
  }

  /**
   * Parses the `config` to an object of {@link DataTableRow}.
   *
   * @param config the config to be parsed.
   * @returns populated {@link DataTableRow} object.
   */
  parseConfig(config: Config): DataTableRow {
    return {
      cells: [
        {
          value: config.subChain.id,
          label: config.subChain.name,
        },
        {
          value: config.baseChain.id,
          label: config.baseChain.name,
        },
      ],
      id: config.id
    };
  }

  /**
   * Parses the list of `configs` and return the sub-chains in the
   * {@link DataTableRowsType} format.
   *
   * @returns `DataTableRowsType` which is read by the data table.
   */
  getConfigsAsDataTableRowsType(): DataTableRowsType {
    return this.configs.map((config: Config) => {
        return this.parseConfig(config);
      }
    );
  }

  render() {
    return (
      <Host>
        <panic-header showMenu={true}/>

        <svc-content-container class={"panic-settings-chains__container"}>
          <svc-surface>
            <svc-label class={"panic-settings-chains__title"}>
              Sub-Chains
            </svc-label>

            <div class={"panic-settings-chains__add-sub-chain-button"}>
              <svc-button
                id={"add"}
                iconName={"add-circle"}
                color={"secondary"}
                href={"/settings/new/sub-chain"}
                onClick={() => localStorage.removeItem(CONFIG_ID_LOCAL_STORAGE_KEY)}
              >
                Add a sub-chain
              </svc-button>
            </div>

            <svc-data-table
              class={"panic-settings-chains__crud-data-table"}
              mode={"crud"}
              cols={SUBCHAIN_TABLE_HEADERS}
              rows={this.getConfigsAsDataTableRowsType()}
              editEventName={"editConfig"}
              deleteEventName={"deleteConfig"}
              noRecordsMessage={"Click in the button above to add your first configuration..."}
            />

          </svc-surface>
        </svc-content-container>

        <panic-footer/>

      </Host>
    );
  }
}
