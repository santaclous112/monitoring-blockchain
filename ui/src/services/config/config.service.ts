import {ServiceInterface} from "../service.interface";
import {BaseSerializer} from "../serializer.interface";
import {ConfigSerializer} from "./config.serializer";
import {Config} from "../../../../entities/ts/Config";
import {API_URL} from "../../utils/constants";
import { HelperAPI } from "../../utils/helpers";

/**
 * Fetches "raw data" from the base chain endpoints and convert them to the
 * equivalent entity model.
 *
 * ConfigService is a singleton. To get the instance of this service, you must
 * call `ConfigService.getInstance()`.
 */
export class ConfigService implements ServiceInterface<Config> {

    ENDPOINT_URL: string = `${API_URL}v1/configs`;
    serializer: BaseSerializer<Config> = new ConfigSerializer();

    /**
     * It represents the `single` instance shared across the consumers of
     * `ConfigService`.
     */
    private static _instance: ConfigService;

    /**
     * Since this class is a `singleton`, we cannot allow object creation.
     */
    private constructor() { }

    /**
     * This static method is used to get an instance of {@link ConfigService}.
     *
     * @returns the `singleton` instance of {@link ConfigService}
     */
    public static getInstance(): ConfigService {
        if (!ConfigService._instance)
            ConfigService._instance = new ConfigService();

        return ConfigService._instance;
    }

    async getAll(): Promise<Config[]> {
        try {
            const configs: Config[] = [];
            const response = await fetch(this.ENDPOINT_URL);
            const json = await response.json();

            json.result.forEach((serializedConfig) => {
                configs.push(serializedConfig);
            });

            return new Promise((resolve) => { resolve(configs); });
        } catch (error) {
            HelperAPI.raiseToast("Could not communicate with the database. Check the API Docker service!", 3000, "danger");
        }
    }

    async getByID(id: string, suppressToast: boolean = false): Promise<Config> {
        try {
            const config = await fetch(`${this.ENDPOINT_URL}/${id}`, { method: 'GET' });
            const json = await config.json();

            return Config.fromJSON(json.result, Config);
        } catch (error) {
            if(!suppressToast){
                HelperAPI.raiseToast("Could not communicate with the database. Check the API Docker service!", 3000, "danger");
            }
        }
    }

    async save(config: Config): Promise<Response> {
        const endpoint: string = config.id ? `${this.ENDPOINT_URL}/${config.id}` : this.ENDPOINT_URL;
        const httpMethod: string = config.id ? 'PUT' : 'POST';

        try {
            return await fetch(endpoint,
                {
                    method: httpMethod,
                    headers: {
                        'Accept': 'application/json',
                        'Content-Type': 'application/json'
                    },
                    body: JSON.stringify(config),
                });
        } catch (error) {
            HelperAPI.raiseToast("Could not communicate with the database. Check the API Docker service!", 3000, "danger");
        }
    }

    async delete(configId: string): Promise<boolean> {
        try {
            const response = await fetch(`${this.ENDPOINT_URL}/${configId}`, { method: "DELETE" });

            return response.ok;
        } catch (error) {
            HelperAPI.raiseToast("Could not communicate with the database. Check the API Docker service!", 3000, "danger");
        }
    }
}
