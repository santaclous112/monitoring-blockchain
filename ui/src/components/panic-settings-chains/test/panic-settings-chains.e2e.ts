import { newSpecPage } from '@stencil/core/testing';
import { PanicSettingsChains } from '../panic-settings-chains';
import {ConfigService} from "../../../services/config/config.service";

const dummyConfigs = [{
    subChain: {name: "test_sub_chain"},
    baseChain: {name: "test_base_chain"},
    channels: [
        {
            "channel_id": 1651233337426,
            "enabled_sub_chains": [
                "test_sub_chain"
            ],
            "channel_data": {
                "channel_type": "telegram",
                "info": "true",
                "warning": "",
                "critical": "true",
                "error": "true",
                "channel_name": "telegram_bot_1",
                "bot_token": "token_123",
                "chat_id": "123",
                "telegram_commands": "true",
                "telegram_alerts": "true"
            }
        },
        {
            "channel_id": 1651233435456,
            "enabled_sub_chains": [],
            "channel_data": {
                "channel_type": "slack",
                "info": "true",
                "warning": "",
                "critical": "",
                "error": "",
                "channel_name": "slack_bot_1",
                "app_token": "app123",
                "bot_token": "123",
                "bot_channel_id": "123123",
                "commands": "",
                "alerts": "true"
            }
        },
    ],
    repos: [],
}]
ConfigService.getInstance().getAll = jest.fn().mockReturnValue(dummyConfigs);

describe('panic-settings-chains', () => {
    it('renders', async () => {
        const page = await newSpecPage({
            components: [PanicSettingsChains],
            html: '<panic-settings-chains></panic-settings-chains>',
        });

        const panicRoot = page.body.querySelector('panic-settings-chains');

        expect(panicRoot).toMatchSnapshot();
    });
});
