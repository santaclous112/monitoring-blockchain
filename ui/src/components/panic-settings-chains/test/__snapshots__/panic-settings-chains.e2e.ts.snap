// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`panic-settings-chains renders 1`] = `
<panic-settings-chains>
  <panic-header showmenu=""></panic-header>
  <svc-content-container class="panic-settings-chains__container">
    <svc-surface>
      <svc-label class="panic-settings-chains__title">
        Sub-Chains
      </svc-label>
      <div class="panic-settings-chains__add-sub-chain-button">
        <svc-button color="secondary" href="/settings/new/sub-chain" iconname="add-circle" id="add">
          Add a sub-chain
        </svc-button>
      </div>
      <svc-data-table class="panic-settings-chains__crud-data-table" deleteeventname="deleteConfig" editeventname="editConfig" mode="crud" norecordsmessage="Click in the button above to add your first configuration..."></svc-data-table>
    </svc-surface>
  </svc-content-container>
  <panic-footer></panic-footer>
</panic-settings-chains>
`;
